File: 0860-lemonade-change/0860-lemonade-change.py
class Solution(object):
    def lemonadeChange(self, bills):
        """
        :type bills: List[int]
        :rtype: bool
        """
        a = 0 
        b = 0 
        c = 0 

        for i in bills:
            
            if i == 5 :
                a+=1
                continue 

            elif i == 10:
                b+=1
                if a > 0:
                    a-=1
                    continue
                else:
                    return False 
            else:
                c+=1 
                if a > 2 or (a > 0 and b > 0):
                    if a > 0 and b > 0:
                        a -= 1 
                        b -= 1
                    else:
                        a -= 3
                        

                    continue
                else:
                    return False

        return True 
        

File: 0860-lemonade-change/0860-lemonade-change.java
class Solution {
    public boolean lemonadeChange(int[] bills) {
        if(bills[0] != 5) return false;;
        
        HashMap<Integer, Integer> map = new HashMap<>();


        for(int i = 0; i < bills.length; i++){
            int rupee = bills[i];
            map.put(rupee, map.getOrDefault(rupee, 0) + 1);
            if(rupee == 10){
                //check 5 rupee - 1
                if(map.getOrDefault(5, 0) >= 1){
                    map.put(5, map.get(5) - 1);
                    if(map.get(5) == 0) map.remove(5);
                }
                else{
                    return false;
                }
            }
            else if(rupee == 20){
                // check 3- 5 rupee or 1-ten rupee and 1- 5 rupee
                if(map.getOrDefault(10, 0) >= 1 && map.getOrDefault(5, 0) >= 1){
                    map.put(10, map.get(10) - 1);
                    map.put(5, map.get(5) - 1);
                    if(map.get(5) == 0) map.remove(5);
                    if(map.get(10) == 0) map.remove(10);
                }
                else if(map.getOrDefault(5, 0) >= 3){
                    map.put(5, map.get(5) - 3);
                    if(map.get(5) == 0) map.remove(5);
                }
                else{
                    return false;
                }

            }
        }
        return true;
    }
}